src/mcts/node.cc

```diff
@@ -32,247 +32,151 @@
-      // Reserve is empty now, so unless another thread did that, we have to
-      // rebuild a new reserve.
-      if (!reserve_list_) AllocateNewBatch();
-    }
-    // Repeat again, now as we have reserve list and (possibly) free list.
-  }
+std::string Edge::DebugString() const {
+  std::ostringstream oss;
+  oss << "Move: " << move_.as_string() << " P:" << p_;
+  return oss.str();
 }
 
-void Node::Pool::AllocateNewBatch() REQUIRES(allocations_mutex_) {
-  allocations_.emplace_back(std::make_unique<FreeNode[]>(kAllocationSize));
+/////////////////////////////////////////////////////////////////////////
+// EdgeList
+/////////////////////////////////////////////////////////////////////////
 
-  FreeNode* new_nodes = allocations_.back().get();
-  for (int i = 0; i < kAllocationSize; ++i) {
-    FreeNode* n = new_nodes + i;
-    n->next = reserve_list_;
-    reserve_list_ = n;
-  }
+EdgeList::EdgeList(MoveList moves)
+    : edges_(static_cast<Edge*>(operator new(sizeof(Edge) * moves.size()))),
+      size_(moves.size()) {
+  auto* edge = edges_;
+  for (auto move : moves) new (edge++) Edge(move);
 }
 
-void Node::Pool::ReleaseChildren(Node* node) {
-  Node* next = node->child_;
-  // Iterating manually rather than with iterator, as node is released in the
-  // middle and can be taken by other threads, so we have to be careful.
-  while (next) {
-    Node* iter = next;
-    // Getting next after releasing node, as otherwise it can be reallocated
-    // and overwritten.
-    next = next->sibling_;
-    ReleaseSubtree(iter);
-  }
-  node->child_ = nullptr;
+void EdgeList::DestroyList() {
+  if (!edges_) return;
+  for (int i = 0; i < size_; ++i) edges_[i].~Edge();
+  operator delete(static_cast<void*>(edges_));
+  edges_ = nullptr;
 }
 
-void Node::Pool::ReleaseAllChildrenExceptOne(Node* root, Node* subtree) {
-  Node* child = nullptr;
-  Node* next = root->child_;
-  while (next) {
-    Node* iter = next;
-    // Getting next after releasing node, as otherwise it can be reallocated
-    // and overwritten.
-    next = next->sibling_;
-    if (iter == subtree) {
-      child = iter;
-    } else {
-      ReleaseSubtree(iter);
-    }
-  }
-  root->child_ = child;
-  if (child) {
-    child->sibling_ = nullptr;
-  }
+void EdgeList::operator=(EdgeList&& other) noexcept {
+  DestroyList();
+  edges_ = other.edges_;
+  size_ = other.size_;
+  other.edges_ = nullptr;
+  other.size_ = 0;
 }
 
-void Node::Pool::ReleaseSubtree(Node* node) {
-  Mutex::Lock lock(gc_mutex_);
-  subtrees_to_gc_.push_back(node);
-}
-
-Node::Pool gNodePool;
-
 /////////////////////////////////////////////////////////////////////////
 // Node
 /////////////////////////////////////////////////////////////////////////
 
-Node* Node::CreateChild(Move m) {
-  Node* new_node = gNodePool.AllocateNode();
-  new_node->parent_ = this;
-  new_node->sibling_ = child_;
-  new_node->move_ = m;
-  child_ = new_node;
-  return new_node;
+Node* Node::CreateSingleChildNode(Move move) {
```

Comment:
This method feels risky (in that the current node ends up in a state which cannot be used in search - so if you ever want to go back to it, you have to reset it completely).  But I've not identified any actual problem.

Reply:
The game tree looks like this:
```
           Node    <- gamebegin_node_
            |
           Node
            |
           Node
            |
           Node
            |
           Node    <- current_head_
  /  / / / | \ \ \ \  \  \  \  \
 Node Node Node Node Node Node Node
 /|\  /|\  /|\  /|\   /|\ /|\  /|\ \ \\ \\ 
Node Node Node Node Node Node Node Node Node
```
I.e. between game_begin and current move it's essentially a linked list, and for the search it's a tree. This allows to have the same code for e.g. `position ... moves a1a2 b1b2` with tree reuse independent on whether it's shorter or longer from previous `position` command. Also it's very handy when filling history planes.

The `CreateSingleChildNode()` is used to build that part of a tree between gamebegin_node_ and current_head_.

Those lines in NodeTree::ResetToPosition() ensure that if new search position is shorter than previous, the search tree is properly trimmed:
```cpp
  // If we didn't see old head, it means that new position is shorter.
  // As we killed the search tree already, trim it to redo the search.
  if (!seen_old_head) {
    assert(!current_head_->sibling_);
    TrimTreeAtHead();
  }
```

I fill that I should add all that somewhere in comments in the code.